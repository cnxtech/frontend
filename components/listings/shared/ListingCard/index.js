import React, { Component } from 'react'
import Router from 'next/router'
import NumberFormat from 'react-number-format'
import Link from 'next/link'
import humps from 'humps'

import {
  buildSlug,
  getListingSummary
} from 'lib/listings'
import LikeButton from 'components/shared/Common/Buttons/Like'
import { thumbnailUrl } from 'utils/image_url'
import { intToCurrency } from 'utils/text-utils'
import Row from '@emcasa/ui-dom/components/Row'
import Col from '@emcasa/ui-dom/components/Col'
import View from '@emcasa/ui-dom/components/View'
import Text from '@emcasa/ui-dom/components/Text'
import Button from '@emcasa/ui-dom/components/Button'
import {
  Container,
  ListingImage,
  LikeButtonContainer,
  THUMB_WIDTH,
  THUMB_HEIGHT
} from './styles'

class ListingCard extends Component {
  handleListingClick = () => {
    const {listing} = this.props
    Router.push(`/listings/show?id=${listing.id}`, buildSlug(listing)).then(
      () => window.scrollTo(0, 0)
    )
  }

  render() {
    let {
      listing,
      currentUser,
      favorited: favoritedListings,
      loading,
    } = this.props
    listing = humps.decamelizeKeys(listing)

    const favorited =
      favoritedListings.filter(
        (actual) => actual.id.toString() === listing.id.toString()
      ).length > 0

    const thumbFilename = listing.images && listing.images[0] ? listing.images[0].filename : ''
    const thumbUrl = thumbnailUrl(thumbFilename, THUMB_WIDTH * 2, THUMB_HEIGHT * 2)

    console.log('listing:', listing)
    const listingSummary = getListingSummary(listing)

    return (
      <Link
        href={`/listings/show?id=${listing.id}`}
        as={buildSlug(listing)}
        passHref
      >
        <Container aria-label={`listing-${listing.id}`}>
          <ListingImage url={thumbUrl} />
          <Row flexDirection="column" p={2}>
            <Row><Text inline fontSize="small">{listing.address.neighborhood.toUpperCase()}</Text></Row>
            <Row><Text inline fontSize="small">{listing.address.street}</Text></Row>
            <Row><Text inline fontSize="small" color="grey">{listingSummary}</Text></Row>
            <Row><Text inline fontSize="large" fontWeight="bold">{intToCurrency(listing.price)}</Text></Row>
          </Row>
          <LikeButtonContainer>
            <LikeButton
              favorite={favorited}
              listing={listing}
              user={currentUser}
              secondary
            />
          </LikeButtonContainer>
        </Container>
      </Link>
    )
  }
}

export default ListingCard
