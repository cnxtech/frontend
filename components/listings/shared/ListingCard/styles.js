import styled from 'styled-components'
import {themeGet} from 'styled-system'
import View from '@emcasa/ui-dom/components/View'
import {
  shouldShowMap,
  MAP_WIDTH_PERCENT
} from 'components/listings/shared/ListingList/styles'

const MIN_CARD_WIDTH = 280
const CARD_MARGIN = themeGet('space.4')

const getCardWidth = (props) => {
  if (!process.browser) {
    return
  }

  const cardMargin = CARD_MARGIN(props)
  const clientWidth = Math.floor(document.documentElement.clientWidth)
  // Two margins (each corner of the document: left of list, right of map)
  const pageMargins = cardMargin * 2

  // Map width to be discounted
  const showMap = shouldShowMap()
  const mapWidth = showMap
    ? Math.ceil(clientWidth * (MAP_WIDTH_PERCENT / 100))
    : 0

  // Calculated area to fit cards (in one row)
  const cardsArea =
    clientWidth - mapWidth - pageMargins + (showMap ? 0 : cardMargin)

  // How many cards, minimum, can fit in this row?
  const cardsPerRow = Math.floor(cardsArea / (MIN_CARD_WIDTH + cardMargin))
  return cardsArea / cardsPerRow - cardMargin
}

const Container = styled(View)`
  position: relative;
  box-sizing: border-box;
  cursor: pointer;
  width: ${(props) => getCardWidth(props)}px;
  height: auto;
  border: 1px solid ${themeGet('colors.lightGrey')};
  border-radius: ${themeGet('space.1')}px;
  margin: 0 ${themeGet('space.4')}px ${themeGet('space.4')}px 0;

  :hover {
    border-color: ${themeGet('colors.pink')};
    box-shadow: 0 0 2px ${themeGet('colors.pink')};
  }

  img {
    box-sizing: border-box;
    object-fit: cover;
    width: 100%;
    min-height: ${(props) => Math.round(getCardWidth(props) * 0.5)}px;
    height: ${(props) => Math.round(getCardWidth(props) * 0.5)}px;
    border-radius: ${themeGet('space.1')}px ${themeGet('space.1')}px 0 0;
  }
`

export {Container, getCardWidth}
