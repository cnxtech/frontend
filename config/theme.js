import theme from '@emcasa/ui'
import {fas} from '@fortawesome/pro-solid-svg-icons'
import {fal} from '@fortawesome/pro-light-svg-icons'
import icons from './icons'

export default {
  ...theme,
  icons: {emcasa: icons, default: fas, light: fal}
}
