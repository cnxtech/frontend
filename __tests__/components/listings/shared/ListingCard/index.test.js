import React from 'react'
import ListingCard from 'components/listings/shared/ListingCard'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import theme from '@emcasa/ui'
import { ThemeProvider } from 'styled-components'
import { MockedProvider } from 'react-apollo/test-utils'

describe('<ListingCard/>', () => {
  it('should render component', () => {
    const tree = renderer.create(
      <MockedProvider>
      <ThemeProvider theme={theme}>
        <ListingCard
          listing={{
            price: 1000000,
            address: {
              city: 'Rio de Janeiro',
              state: 'RJ',
              neighborhood: 'Bairro',
              street: 'Rua tal'
            },
            images: []
          }}
          favorited={[]}
        />
      </ThemeProvider></MockedProvider>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
