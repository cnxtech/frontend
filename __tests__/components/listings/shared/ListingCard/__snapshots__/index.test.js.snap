// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListingCard/> should render component 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 4px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  -webkit-font-smoothing: antialiased;
  font-family: Rubik;
  color: #3c4858;
  font-size: 14px;
  font-weight: 500;
}

.c5 {
  -webkit-font-smoothing: antialiased;
  font-family: Rubik;
  color: #f50057;
  font-size: 14px;
}

.c6 {
  -webkit-font-smoothing: antialiased;
  font-family: Rubik;
  color: #3c4858;
  font-size: 14px;
}

.c0 {
  position: relative;
  width: 100%;
}

.c0 > button {
  z-index: 2;
}

.c0 > button svg {
  -webkit-transition: -webkit-transform .45s cubic-bezier(.4,.2,0,1);
  -webkit-transition: transform .45s cubic-bezier(.4,.2,0,1);
  transition: transform .45s cubic-bezier(.4,.2,0,1);
}

.c0 > button:hover svg {
  -webkit-transform: scale(1.12);
  -ms-transform: scale(1.12);
  transform: scale(1.12);
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  cursor: pointer;
  width: 100%;
  overflow: hidden;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  border: 1px solid #e0e6ed;
  border-radius: 10px;
  background-color: #ffffff;
  -webkit-transition: border .25s,box-shadow .25s ease-in;
  transition: border .25s,box-shadow .25s ease-in;
}

.c1:hover {
  border-color: #f50057;
  box-shadow: 0 0 2px #f50057;
}

.c1 img {
  box-sizing: border-box;
  object-fit: cover;
  width: 100%;
  height: 140px;
}

.c7 {
  outline: none;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-color: #e0e6ed;
  background-color: #ffffff;
  border-radius: 4px;
  border-width: 1px;
  width: auto;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  height: 40px;
  color: #3c4858;
  font-family: Rubik;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-style: solid;
  position: absolute;
  cursor: pointer;
  right: 14px;
  top: 90px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  box-shadow: 0 6px 16px 0 rgba(38,38,38,0.15);
}

.c7 svg {
  width: 25px;
  text-shadow: 2px 2px 3px #f00;
  text-shadow: 2px 2px 3px #f00;
}

.c7 svg path {
  text-shadow: 2px 2px 3px #f00;
  fill: #ffffff;
  stroke: #3c4858;
  fill-opacity: 0;
  stroke-width: 60;
}

<article
  className="c0"
>
  <a
    aria-label="listing-undefined"
    className="c1"
    href="/imoveis/rj/rio-de-janeiro/bairro/rua-tal/id-undefined"
    onClick={[Function]}
  >
    <img
      alt="Imagem do imÃ³vel ID-undefined na Rua tal, Bairro, Rio de Janeiro - RJ"
      decoding="async"
      src="https://res.cloudinary.com/emcasa/image/upload/f_auto,c_fit,h_600,w_600/v1513818385/"
    />
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <span
          className="c4"
          color="dark"
          fontSize="small"
          fontWeight="bold"
        >
          Bairro
        </span>
        <span
          className="c5"
          color="pink"
          fontSize="small"
        >
          R$1,000,000
        </span>
      </div>
      <span
        className="c6"
        color="dark"
        fontSize="small"
      >
        Rua tal
      </span>
      <span
        className="c6"
        color="dark"
        fontSize="small"
      >
        
      </span>
    </div>
  </a>
  <button
    className="c7"
    fontSize="default"
    height="medium"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-heart fa-w-16 fa-1x "
      data-icon="heart"
      data-prefix="fas"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</article>
`;
